### DISCUSSION TITLE
Plugin integration and scaffolding tools

### DISCUSSION THREAD

**Discussion by joshuatf** [woocommerce/woocommerce#40049] [General]
## Goals for integration extensibility

Setting up third party plugins and even internal plugins can be challenging, especially when blocks and modern JS build environments are needed.

While the newer environments offer arguably better development environments, the time to accomplish simple extensibility tasks increases greatly.  We should aim to make this a simpler process all around and ease the barrier to entry for developers who are not as familiar with these tools.

In short, the goals should be:

* [ ] Adding scripts to assist the build process into existing plugins
* [ ] Attempting to warn against potential problems in a plugin that may hinder development
* [ ] Copying template files to a plugin with its respective namespacing
* [ ] Provide concrete working examples of extensibility into various area of WooCommerce
* [ ] Allowing easy integration of copied files into the existing plugin
* [ ] Allow for integration of multiple examples into a plugin

## Existing tools

There are a handful of tools and attempts that been created at solving this problem or very similar problems.  The following are all great tools though they don't extensively integrate into an existing plugin and mostly work around creating a new plugin.

### WooCommerce Admin Example Extensions

[This was our original attempt](https://github.com/woocommerce/woocommerce/tree/2fae3537a786a967877683a5635835fe510e3dec/plugins/woocommerce-admin/docs/examples/extensions) at helping plugin authors get started in more React-focused work.  It worked by simply copying over an example plugin file and build process to the plugins directory.

This created a working plugin with a focus on helping extend a specific area of the admin interface, but did not offer much in the way of customization.

### @woocommerce/create-woo-extension

[This is the next iteration](https://github.com/woocommerce/woocommerce/blob/a995cce7e6a99116320d0d906d618c773d109c41/packages/js/create-woo-extension/README.md?plain=1) of the admin extension script.  It is very similar to the above, but piggy backs on `@wordpress/create-block` which gives it the addition of some built-in templating abilities.

### @wordpress/create-block

[This is a fantastic tool in Gutenberg](https://github.com/WordPress/gutenberg/blob/14600f31582ba0ebe7343130d864ac6dba1d73db/packages/create-block/README.md?plain=1) that helps scaffold a block plugin.  It's fairly customizable and comes with a number of options, including the ability to pass a custom template config.

Not only does this tool allow template files and create a configurable build environment, it also has an option `--no-plugin` which scaffolds a block into an existing plugin.  The only downside is that this process is very focused on the creation of a single generic block (in case the package name didn't give it away) and skipping plugin creation also skips scaffolding any plugin related files that may be used to help register and initialize features.

Also worth noting that @mattsherman made a [great template for the product block editor](https://github.com/woocommerce/woocommerce/blob/a995cce7e6a99116320d0d906d618c773d109c41/packages/js/create-product-editor-block/README.md?plain=1) that utilizes `create-block` and works well to set up a working product block, but faces the same challenges with existing plugins.

## Integrative concept

I've been working on a [POC](https://github.com/woocommerce/woocommerce/pull/39941 ) that takes the best of existing work and adapts it to work with existing plugins.  This work is heavily inspired by that done in `@wordpress/create-block`.

### What this script does

When `npx woo-integrate-plugin` is run by itself:

* Initializes a `package.json` file if one does not exist, pulling information from the plugin's metadata
* Initializes a `composer.json` file if one does not exist, again puling from plugin metadata
* Installs the necessary `@wordpress/scripts` if the option is passed, allowing blocks to be easily parsed during the build process
* Adds a number of helper scripts to the `package.json` file, such as `npm run build`.  If a conflicting script key already exists, the developer is warned during the process
* Audits the webpack config, if one exists, to make sure that the WP default config is being used and that external dependencies are being used
* Installs `@wordpress/env` to the packages if the option is passed and adds a script to start the environment.

If a template argument is also provided `npx woo-integrate-plugin --template=...`:

* Copies client-side files from templates to a provided directory, e.g. `src/`
* Copies server-side files from templates to a provided directory, e.g., `includes/`
* Copies plugin-related files to the root director
* Updates a configuration file `.woo-plugin.json` with installed modules to track entry points for these extensions
* Provides the developer with instructions on lines to copy/paste to initialize the template within their existing plugin entry file

<img width="1314" alt="Screen Shot 2023-08-31 at 5 18 44 PM" src="https://github.com/woocommerce/woocommerce/assets/10561050/d90016f8-de8a-4530-95df-d1c1e6722740">

### Template structure

The example in the POC does a decent job of portraying template structure, but ultimately I think files break down into a few buckets.  These could certainly be broken down to be even more granular, but I think this is adequate for most plugins.

* Back-end files (php)
* Client-side files (js/css/assets)
* Blocks (typically a directory under client-side files)
* Entry points (commonly shared)
* Plugin files

While not explicitly related to the integration scripts, I think it's worth noting that the entry points are key in this concept for 3 reasons:

1. They are an easy, single point of entry to detect and load the module using our configuration.  E.g., `BlockRegistry` is the single entry point that loads all registered blocks.
2. They can be shared across templates.  E.g., `BlockRegistry` is likely to be used as an entry point for multiple templates
3. They provide one-liners for plugin developers to add to their own plugins.  E.g., `BlockRegistry::init();`

## Thoughts and discussion

When I started on this concept, I initially thought it may be a fruitless effort and not considerably more useful than providing a set of instructions to help developers set up plugins.  But as it progressed and seeing how many things need to be added or even just audited before a plugin can work, I think this may be a valuable tool.  If this is a project we want to move forward with, I think there are a few things that need to be hashed out before implementation:

* **Overwrite vs warn** - In many cases conflicting files may exist, either from the plugin itself or across templates.  In these cases do we simply overwrite them and allow developers to handle issues via version control (how it's currently done in `create-block`)?  Or do we warn against potential overwrites and provide options?
* **Shared template files** - Mentioned above under "Template structure"; I'd love to see these entry points somehow shared across templates as opposed to copying the same file into every template.
* **Dependency conflicts** - There will be conflicts with existing engines, packages, or dependencies at some point.  Can we come up with adequate ways of gracefully handling or at least warning about these conflicts?  I think so, but are there any cases that stand out as troublesome?
* **Entry points** - The current POC prompts the user to add a single line to load the module entry point from a template.  Are there options to make this work directly out of the box without any manual intervention?  So far, I have not found a "safe" way of doing this, but injection into the top of the plugin's entry file (right after the metadata) may be a candidate.

Are there other areas I haven't considered that add challenges to this approach?  Certain types of plugins that could prove especially difficult?

**Comment by mattsherman** [comment]
> **Overwrite vs warn** - In many cases conflicting files may exist, either from the plugin itself or across templates. In these cases do we simply overwrite them and allow developers to handle issues via version control (how it's currently done in `create-block`)? Or do we warn against potential overwrites and provide options?

I think we should definitely warn against potential overwrites, because the developer may have made edits to the files that they haven't yet committed, staged, or stashed.

**Reply by joshuatf** [reply]
That's a great point, @mattsherman.  That then brings up the question of how to deal with module entry points which might be shared across multiple templates (e.g., `BlockRegistry` in the POC).

One option would be to check if the file is identical and we can skip it if so.  But this might also be alleviated if we come up with a better system for template entry points that are more easily shared across templates.

**Comment by joshuatf** [comment]
cc @louwie17 for visibility.  If you have any questions prior to our call, feel free to post here.  Otherwise I'll recap any findings after we discuss.
